Documentation/teaching/labs/device_drivers.rst

```diff
@@ -948,46 +948,82 @@ For this:
    1. Implement the ioctl function in the driver.
    2. We need to use ``user/so2_cdev_test.c`` to call the
       ioctl function with the appropriate parameters.
+   3. To test, we will use an user-space program (``user/so2_cdev_test.c``)
+      which will call the ``ioctl`` function with the required arguments.
+
+.. note:: The macro ``MY_IOCTL_PRINT`` is defined in the file ``include/so2_cdev.h``,
+          which is shared between the kernel module and the user-space program.
 
-.. note:: The macro definition ``MY_IOCTL_PRINT`` is defined in the ``include/so2_cdev.h`` file
           Read the `ioctl`_ section in the lab.
 
-.. note:: Because we need to compile the program for qemu machine which is 32 bit, if your host is 64 bit
-          then you need to install ``gcc-multilib`` package.
+.. note:: The userspace code is compiled automatically at ``make build`` and
+          copied at ``make copy``.
+
+          Because we need to compile the program for qemu machine which is 32 bit,
+          if your host is 64 bit then you need to install ``gcc-multilib`` package.
+
+Extra Exercises
+===============
+
+Ioctl with messaging
+--------------------
+
+Add two ioctl operations to modify the message associated with the
+driver. Use fixed-length buffer ( BUFFER_SIZE ).
+
+  1. Add the ``ioctl`` function from the driver the following operations:
+     * ``MY_IOCTL_SET_BUFFER`` for writing a message to the device;
+     * ``MY_IOCTL_GET_BUFFER`` to read a message from your device.
+  2. Change the user-space program to allow testing.
+
+.. note:: Read the ioctl sections and Access to the address space of the lab process.
```

Comment:
Don't forget to add hyperlinks to the sections.

Reply:
Fixed.